File: hooks/tts-timer-start.py
#!/usr/bin/env python3
"""Cross-platform hook wrapper: tts-timer-start.

On Unix: delegates to tts-timer-start.sh.
On Windows: native Python implementation.

Claude Code Hook Protocol (PreToolUse):
  Receives JSON on stdin: {"tool_name": "...", "tool_input": {...}, ...}
  Exit 0: always (timing hook, never blocks)

FAILURE POLICY: FAIL-OPEN
"""
import json
import os
import sys
import tempfile
import time
from pathlib import Path

HOOK_DIR = Path(__file__).parent


def main() -> None:
    if sys.platform != "win32":
        script = HOOK_DIR / "tts-timer-start.sh"
        os.execv("/bin/bash", ["bash", str(script)] + sys.argv[1:])

    # Windows: native Python implementation
    stdin_data = sys.stdin.read()
    if not stdin_data:
        sys.exit(0)

    try:
        data = json.loads(stdin_data)
    except json.JSONDecodeError:
        sys.exit(0)

    tool_use_id = data.get("tool_use_id", "")
    if not tool_use_id:
        sys.exit(0)

    try:
        start_file = Path(tempfile.gettempdir()) / f"claude-tool-start-{tool_use_id}"
        start_file.write_text(str(int(time.time())))
    except OSError:
        pass  # Fail-open

    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/nim/src/tts_timer_start.nim
## tts_timer_start - PreToolUse hook for recording tool start times.
##
## Phase: PreToolUse (catch-all, no matcher)
## Timeout: 5s, async
## Failure policy: FAIL-OPEN (timing failures never prevent tool execution)
##
## Writes current Unix timestamp to /tmp/claude-tool-start-{tool_use_id}.
## The companion tts_notify PostToolUse hook reads and deletes this file.

import std/[json, os, strutils]
import hooklib

proc main() =
  let input = readStdinJson()
  if input.isNil: quit(0)

  let toolUseId = input.getOrDefault("tool_use_id").getStr("")
  let sanitized = sanitizeToolUseId(toolUseId)
  if sanitized.isNone: quit(0)

  let ts = $unixTimestamp()
  let path = "/tmp/claude-tool-start-" & sanitized.get
  try:
    writeFile(path, ts & "\n")
  except IOError:
    discard  # fail-open

  quit(0)

main()
